components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java

```diff
@@ -0,0 +1,113 @@
+import org.apache.oltu.oauth2.common.exception.OAuthSystemException;
+import org.apache.oltu.oauth2.common.message.OAuthResponse;
+import org.json.JSONObject;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.device.codegenerator.GenerateKeys;
+import org.wso2.carbon.identity.oauth2.device.constants.Constants;
+import org.wso2.carbon.identity.oauth2.device.dao.DeviceFlowPersistenceFactory;
+import org.wso2.carbon.identity.oauth2.device.errorcodes.DeviceErrorCodes;
+
+import java.util.UUID;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+
+@Path("/device_authorize")
+public class DeviceEndpoint {
+
+    private static final Log log = LogFactory.getLog(DeviceEndpoint.class);
+
+    @POST
+    @Path("/")
+    @Consumes("application/x-www-form-urlencoded")
+    @Produces("application/json")
+    public Response authorize(@Context HttpServletRequest request, @Context HttpServletResponse response)
+            throws IdentityOAuth2Exception {
+
+        String clientId = request.getParameter(Constants.CLIENT_ID);
+        JSONObject errorResponse = new JSONObject();
+        if (StringUtils.isBlank(clientId)) {
+            errorResponse.put(Constants.ERROR, DeviceErrorCodes.INVALID_REQUEST)
+                    .put(Constants.ERROR_DESCRIPTION, "Request missing required parameters");
+            Response.ResponseBuilder respBuilder = Response.status(HttpServletResponse.SC_BAD_REQUEST);
+            return respBuilder.entity(errorResponse.toString()).build();
+        }
+        if (!validateClientId(clientId)) {
+            errorResponse.put(Constants.ERROR, DeviceErrorCodes.UNAUTHORIZED_CLIENT)
+                    .put(Constants.ERROR_DESCRIPTION, "No registered client with the client id.");
+            Response.ResponseBuilder respBuilder = Response.status(HttpServletResponse.SC_UNAUTHORIZED);
+            return respBuilder.entity(errorResponse.toString()).build();
+        }
+        String userCode = GenerateKeys.getKey(Constants.KEY_LENGTH);
+        String deviceCode = UUID.randomUUID().toString();
+        String scope = request.getParameter(Constants.SCOPE);
+        String redirectionUri = IdentityUtil.getServerURL("/authenticationendpoint/device.do", false, false);
+        String redirectionUriComplete = redirectionUri + "?user_code=" + userCode;
+        DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO().insertDeviceFlowParameters(deviceCode,
+                userCode, clientId, Constants.EXPIRES_IN_VALUE, Constants.INTERVAL_VALUE);
+        DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO().storeDeviceFlowScopes(scope, deviceCode);
```

Comment:
1. Do not invoke, the DAO layer from the endpoint. Introduce a service layer to obtain the the DAO
2. Inserting device flow parms and scopes should be atomic (should be done in a single transaction)